enhancements/etcd/protecting-etcd-quorum-during-control-plane-scaling.md

```diff
@@ -0,0 +1,452 @@
+    after a new member joins the cluster
+  * This also ensures that we keep the full etcd data on disk on a minimum of the desired number of Control Plane
+    Machines at all times
+
+#### Protecting etcd during scaling operations
+
+To ensure that the safety requirements described above are maintained during scaling operations,
+the etcd operator will manage the etcd cluster in clusters with a
+[functional Machine API](#When-is-Machine-API-Functional),
+by leveraging [Machine Deletion Hooks](#What-are-Machine-Deletion-Hooks) to coordinate with the Machine API when it is,
+and isn't safe, to remove Machines with voting members of the etcd cluster running on them.
+
+##### Overview of protecion mechanism
+
+To ensure the safety of the etcd cluster quorum, the etcd operator will leverage a pre-drain [Machine Deletion Hooks](#What-are-Machine-Deletion-Hook) to prevent the removal of any Control Plane Machine hosting a voting member of the etcd quorum.
+
+```yaml
+lifecycleHooks:
+  preDrain:
+  - name: EtcdQuorumOperator
+    owner: clusteroperator/etcd
+```
+
+The etcd operator will apply the hook to a Machine resource once it identifies that the Machine hosts an etcd member.
+The hook should be added before the member is promoted to ensure that there is no period where the member is a voting
+member, while the machine is not protected by the deletion mechanism.
+
+It will only remove the hook once it has identified that the Machine resource is being deleted and a replacement member
+has been created. The removal of this hook will allow the Machine API to drain and terminate the Machine as it would
+normally do.
+
+In the case that a Machine is deleted before the member is promoted, the etcd operator is expected to not promote the
+new member, and remove the deletion hook to allow the Machine to be removed from the cluster.
+Once a Machine has been marked for deletion, if the hook is removed by some other entity, the etcd operator
+is expected not to re-add the hook. This allows an escape hatch when manual intervention is required.
+
+The etcd operator will ensure, based on the desired Control Plane replica count in the cluster `InstallConfig`
+resource, that the etcd cluster has either the exact desired count of voting members, or during scaling/replacement
+operations, at most 1 extra voting member.
+
+The etcd operator will also leverage the etcd quorum guard to prevent voluntary disruptions of etcd members during the
+process. By ensuring that the quorum guard PDB always has `minAvailable: (Num Current Control Plane Machines) - 1`,
+this prevents draining of a healthy etcd member until a new member becomes healthy. This operation will prevent other
+components in the cluster (eg. MCO) from disrupting the quorum of etcd during this operation.
+
+Note: When the cluster is already degraded, the etcd operator is expected to report as degraded for admin intervention.
```

Comment:
is there a way to determine a cluster is in a degraded condition?

Reply:
We do have health checks that determine when members are unhealthy
https://github.com/openshift/cluster-etcd-operator/blob/14fecbd21d47fe19ab0219f59a10a0bbd4dc67ea/pkg/etcdcli/health.go#L42

And today we abort membership changes when that's the case
https://github.com/openshift/cluster-etcd-operator/blob/master/pkg/operator/clustermembercontroller/clustermembercontroller.go#L87-L94